src/Transports/MassTransit.AmazonSqsTransport/AmazonSqsTransport/AmazonSqsReceiveLockContext.cs

```diff
@@ -0,0 +1,157 @@
+namespace MassTransit.AmazonSqsTransport
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Amazon.SQS;
+    using Amazon.SQS.Model;
+    using Internals;
+    using Transports;
+
+
+    public class AmazonSqsReceiveLockContext :
+        ReceiveLockContext,
+        IAsyncDisposable
+    {
+        static readonly TimeSpan MaxVisibilityTimeout = TimeSpan.FromHours(12);
+        readonly CancellationTokenSource _activeTokenSource;
+        readonly ClientContext _clientContext;
+        readonly Message _message;
+
+        readonly ReceiveContext _receiveContext;
+        readonly ReceiveSettings _settings;
+        readonly Task<Task> _visibilityTask;
+        bool _locked;
+
+        public AmazonSqsReceiveLockContext(ReceiveContext receiveContext, Message message, ReceiveSettings settings, ClientContext clientContext)
+        {
+            _receiveContext = receiveContext;
+            _message = message;
+            _settings = settings;
+            _clientContext = clientContext;
+            _activeTokenSource = new CancellationTokenSource();
+            _locked = true;
+
+            _visibilityTask = Task.Factory.StartNew(RenewMessageVisibility, _activeTokenSource.Token, TaskCreationOptions.None, TaskScheduler.Default);
+        }
+
+        public async ValueTask DisposeAsync()
+        {
+            _activeTokenSource.Cancel();
+            await _visibilityTask.ConfigureAwait(false);
```

Comment:
If this is the only reason, there is no reason to wait for this task to exit at the expense of everything having to be DisposeAsync(), that's way too much overhead.